lib/environment/background/redirection.js

```diff
@@ -0,0 +1,114 @@
+/* @flow */
+
+import once from 'lodash/once';
+import keyBy from 'lodash/keyBy';
+
+import type { RedirectRule } from '../utils/redirection';
+import { requiredPermissions } from '../utils/redirection';
+import { _get as get, _set as set } from '../utils/storage';
+import { addListener } from './messaging';
+import { checkPermissionsGranted } from './permissions';
+
+let redirectRules: { [string]: RedirectRule };
+
+init();
+
+async function init() {
+	// pick up rules from storage so redirects work on cold start
+	redirectRules = deserializeRedirectRules(await get('RESRedirect', []));
+
+	addListener('redirection', messageListener);
+
+	checkPermissionsGranted(requiredPermissions)
+		.then(granted => {
+			if (granted) addWebReqListener();
+		});
+}
+
+function serializeRedirectRules(rules) {
+	return Object.values(rules).map(rule => {
+		rule.from = rule.from.toString();
+		return rule;
+	});
+}
+
+function deserializeRedirectRules(rules: Array<RedirectRule>) {
+	const deserializedRedirectRules = rules
+		.map(rule => {
+			const deserializedRule = { ...rule };
+			if (typeof deserializedRule.from === 'string') {
+				deserializedRule.from = reviveRegExp(deserializedRule.from);
+			}
+			return deserializedRule.from === null ? null : deserializedRule;
+		})
+		.filter(rule => rule !== null);
+
+	return keyBy(deserializedRedirectRules, x => x.name);
+}
+
+// this listener handles changes to redirect state
+function messageListener(rules: Array<RedirectRule>) {
```

Comment:
We're only storing the rules as an object to make deletions easier. I'm not sure we should require other scripts to be aware of the way the redirect bg script represents redirect rules in its internal state (instead of thinking of them as a generic "bag"/"list" of rules). That sounds like a leaky abstraction.